paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/DynamicPartitionLevelLoader.java

```diff
@@ -0,0 +1,195 @@
+                return newPartitions;
+            } else {
+                return newPartitions.subList(0, 1);
+            }
+        }
+
+        for (int level = 0; level <= maxPartitionLoadLevel; level++) {
+            newPartitions = extractMaxPartitionsForFixedLevel(newPartitions, level);
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(
+                        "DynamicPartitionLevelLoader(currentPartitionLoadLevel={},table={}) finds new partitions: {}.",
+                        level,
+                        table.name(),
+                        partitionsToString(newPartitions));
+            }
+        }
+        return newPartitions;
+    }
+
+    private int getMaxPartitionLoadLevel(
+            Map<String, String> partitionLoadConfig, List<String> partitionFields) {
+        Preconditions.checkArgument(partitionLoadConfig.size() <= partitionFields.size());
+
+        int maxLoadLevel = -1;
+        for (int i = 0; i < partitionFields.size(); i++) {
+            String config = partitionLoadConfig.getOrDefault(partitionFields.get(i), null);
+
+            if (i == 0) {
+                Preconditions.checkArgument(
+                        config != null, "the top level partition must set load config.");
+            }
+            if (config != null) {
+                Preconditions.checkArgument(
+                        config.equals(MAX_PT),
+                        "unsupported load config '{}' for partition field '{}', currently only support '{}'.",
+                        config,
+                        partitionFields.get(i),
+                        MAX_PT);
+                maxLoadLevel = Math.max(maxLoadLevel, i);
+
+            } else {
+                // if level-i partition don't set config, we won't consider the partition level
+                // which is greater than level-i
+                return maxLoadLevel;
+            }
+        }
+        return maxLoadLevel;
+    }
+
+    private List<InternalRow.FieldGetter> createPartitionFieldGetters() {
+        List<InternalRow.FieldGetter> fieldGetters = new ArrayList<>();
+
+        RowType partitionType = table.rowType().project(table.partitionKeys());
+
+        for (int i = 0; i < maxPartitionLoadLevel + 1; i++) {
+            fieldGetters.add(InternalRow.createFieldGetter(partitionType.getTypeAt(i), i));
+        }
+        return fieldGetters;
+    }
+
+    private List<BinaryRow> extractMaxPartitionsForFixedLevel(
+            List<BinaryRow> partitions, int level) {
+        AtomicInteger currentDistinct = new AtomicInteger(0);
+        Object lastField = null;
+        for (int i = 0; i < partitions.size(); i++) {
+            BinaryRow partition = partitions.get(i);
+            Object newField = fieldGetters.get(level).getFieldOrNull(partition);
+            // if newField is null, it's the default partition
+            if (newField == null) {
+                newField = "__DEFAULT_PARTITION__";
```

Comment:
Might be better to introduce a constant for `"__DEFAULT_PARTITION__"`. Or use `CoreOptions.PARTITION_DEFAULT_NAME.defaultValue()` here.